### E-Klausur "Wissensrepräsentation" - 1. Prüfungszeitraum

Datum: 13. Juli 2021    
Zeit: 9:45 - 11:15

Schreiben Sie zuerst Ihre Namen und ihre Matrikelnummer in das Notebook.


Name: 

Matrikelnummer:

Bei der von mir eingereichten Arbeit handelt es sich um eine eKlausur auf Distanz. Die eKlausur wird open book durchgeführt. Das bedeutet, dass öffentlich zugängliche Materialien (wie Internetquellen, Bücher, Zeitschriften) und persönliche Aufzeichnungen verwenden werden dürfen. 

Durch die Abgabe versichere ich, dass ich die von mir eingereichte Arbeit selbstständig ohne fremde Hilfe verfasst zu haben und dass ich prüfungsfähig bin.

#### Hinweise

- Die Zeit für die Bearbeitung der Klausur ist extra sehr knapp bemessen. Diese Vorgabe 
wird gemacht, damit möglichst wenig Zeit für Kommunikation zwischen Studierenden und Recherche im Internet bleibt.

- Bei teilweise identischen Lösungen zwischen Studierenden muss von einem Betrugsversuch ausgegangen werden. Die Strafen hierfür gehen von einer Bewertung mit 5 (nicht-bestanden) bis hin zur Exmatrikulation.

- Eine zu spät abgegebene Klausur muss mit einer 5 (nicht-bestanden) gewertet werden. Laden Sie die Klausur daher rechtzeitig in moodle hoch.

- Formeln sind im Latexstil zu setzen (in doppelten \$\$), z.B.

$$
A = \frac{x^2}{\sqrt{y}}
$$
- Nicht-korrekt gesetzte Formeln können nicht gewertet werden.
- Alternativ kann die Lösung auf eingescanntes DIN A4-Blatt geschrieben werden. Schreiben Sie sauber und deutlich. Nicht lesbare Lösungen (auch schlechter Scan, schlechtes Foto) können nicht gewertet werden.
- Alle Rechenwege sind anzugeben. Ihre Lösungen müssen nachvollziehbar und klar sein.
 Nicht kreuz und quer schreiben! Nicht nachvollziehbare Lösungen führen zu Punktabzug bzw. Nicht-Wertung.
- Wenn Zahlenwerte in der Aufgabenstellung angegeben sind, müssen Sie
diese verwenden und ein konkretes Ergebnis (Zahlenwert) produzieren.

In [12]:
#Ändern Sie nicht diese Zelle 
import numpy as np
 

num_aufgaben = 7
aufgaben_punkte = np.ndarray(num_aufgaben)


In [2]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für Aufgabe 1
aufgabe = 1
aufgaben_punkte[aufgabe -1] = 14



### 1. Aufgabe: Kostenfunktion und Gradient



Gegeben ist das lineare Modell

$$
h_\theta(x) = \theta_0 + \theta_1 x
$$

und folgende Trainingsdaten:

$$
D_{train} = \{(-2,4),(0,0),(1,0),(2,-3)\}
$$

mit den Tuplen $(x , y $: 
$x$ ist das Merkmal und $y$ der Zielwert.


Für die Startwerte $\theta_0 = 1$ und $\theta_1 = -2$ berechnen Sie die folgenden Unteraufgaben:

#### 1.1. Kosten

Berechnen Sie die Kosten für das obige Beispiel
$$ J_D(\theta_0, \theta_1)=\frac{1}{2m}\sum_{i=1}^{m}{(h_\theta(x_i)-y_i)^2} $$

mit den Werten aller Trainingsdaten.

**Solution:**


$ J_D(\theta_0, \theta_1)=\frac{1}{2m}\sum_{i=1}^{m}{(h_\theta(x_i)-y_i)^2} = \frac{1}{2\cdot4}\cdot((1)^2+(1)^2+(-1)^2+(-0)^2)=\frac{3}{8}$

#### 1.2. Gradient

Berechnen Sie die beiden partiellen Ableitungen 
$$ \frac{\partial J (\theta_0, \theta_1)}{\partial \theta_0} $$
und
$$ \frac{\partial J (\theta_0, \theta_1)}{\partial \theta_1} $$


mit den Werten aller Trainingsdaten.

**Solution**

$$ \frac{\partial J (\theta_0, \theta_1)}{\partial \theta_0}=\frac{1}{m}\sum_{i=1}^{m}(\theta_0 + \theta_1 x^i - y^i)=\frac{1}{4}\cdot((1)+(1)+(-1)+(0))=\frac{1}{4}$$

$$\frac{\partial J (\theta_0, \theta_1)}{\partial \theta_1}=\frac{1}{m}\sum_{i=1}^{m}((\theta_0 + \theta_1 x^i - y^i)\cdot x_i)=\frac{1}{4}\cdot (-2+0+(-1)+(0))=-\frac{3}{4}$$



#### 1.3 Gradientenabstiegsschritt

Berechnen Sie die Parameterwerte $\theta_0$ und $\theta_1$ nach dem ersten Gradientenabstiegsschritt mit der Lernrate $\alpha = 0.01$.

**Solution**

$$\theta_{j}^{new} \leftarrow \theta_{j}^{new} - \alpha * \frac{\partial}{\partial\theta_{j}} J(\theta^{old})$$

$$\theta_{0}=1 - 0.01 \cdot \frac{1}{4}=0.9975$$

$$\theta_{1}=-2 - 0.01 \cdot -\frac{3}{4}=-1.9925$$

In [3]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für Aufgabe 2
aufgabe = 2
aufgaben_punkte[aufgabe -1] = 5

### Aufgabe 2

Diskutieren Sie folgende Fragestellung. 

Sie erhalten auf den Trainingsdaten die Genauigkeit (Accuracy) von 0.9 und auf den Validierungsdaten eine 
Genauigkeit von 0.86 für ein festes Modell (z.B. Standardeinstellungen eines scikit-learn Klassifizierer, 
nach völlständigem Training).

Für Ihre Anwendung wollen Sie eine Genauigkeit von 0.92 erzielen.

Ist davon auszugehen, dass mit mehr Trainingsdaten die gewünschte Genauigkeit 
von 0.92 erreicht werden kann? Begründen Sie Ihre Entscheidung.

 


Ja, weil durch neue Traingsdaten dem Model neue Informationen hinzugefügt werden, wenn sie kein rauschen besitzen.

Durch dem hinzufügen neuer Daten und der dadurch erzeugten Diversität reduzieren sich die Wahrscheinlichkeiten auf overfitting.

In [3]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für 
aufgabe = 3
aufgaben_punkte[aufgabe -1] = 4

### 3. Aufgabe: Implementierung des Gradientenabstiegsverfahrens

Folgender Code wird benutzt bei der Implementation des Gradientenabstiegsverfahren 
bei der univariaten lineare Regression:
    
```for i in range(nb_steps):
        theta[1] = theta[1] - alpha * compute_gradient_cost(theta, 1)
        theta[0] = theta[0] - alpha * compute_gradient_cost(theta, 0)```


Dabei besteht `theta` aus den beiden Komponenten `theta[0]` und `theta[1]`. 
Falls z.B. `theta[1]` geändert wird, ändert sich die entsprechende Komponente von `theta`.

`compute_gradient_cost(theta, x)` 
berechnet die partielle Ableitung der
Kostenfunktion nach der entsprechenden Komponente 0 oder 1.


Welcher prinzipielle Fehler ist hier vorhanden? Korrigieren Sie den (pseudo) Code.
 

**Solution**

`for i in range(nb_steps):
        theta[0] = theta[0] - alpha * compute_gradient_cost(theta, 0)
        theta[1] = theta[1] - alpha * compute_gradient_cost(theta, 1)
       `


In [5]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für
aufgabe = 4
aufgaben_punkte[aufgabe -1] = 10

### 4. Aufgabe

Sind die Aussagen richtig? Kommentieren und korrigieren Sie gegebenenfalls die Aussagen. 
Keine Punkte ohne Begründung!


Folgende Aussagen beziehen sich auf die lineare bzw. logistische Regression, wie sie in der Veranstaltung besprochen 
wurden:


#### 4.1. 
Je höher die Lernrate $\alpha$ ist, desto langsamer findet das Gradientenabstiegsverfahren das Minimum der Kostenfunktion.

**Solution**

Nein, desto schneller findet das Gradientenabstiegsverfahren das Minimum der Kostenfunktion (wenn es gefunden wird bei zu großen Schritten).

#### 4.2 
Falls die Lernrate zu groß gewählt wird, divergiert das Gradientenabstiegsverfahren, d.h. der Kostenwert geht mit der Anzahl der Iterationen gegen Unendlich.

**Solution**

Ja, weil durch die großen Schritte kein Minimum gefunden werden kann.

#### 4.3 

Die Lernrate muss während des Trainings erniedrigt werden, damit das Gradientenabstiegsverfahren konvergieren kann.

**Solution**

Nein, es ist auch möglich mit einer gleichbleibenden Lernrate (Anfangskonfiguration) konvergieren des Gradientenabstiegsverfahren zu erzielen.

#### 4.4

Falls das Gradientenabstiegsverfahren konvergiert, bestimmt die Lernrate $\alpha$ die Lage des Minimums im Parameterraum.

**Solution**

Ja, die Lernrate ist dafür verantwortlich wie weit zwischen den Schritten des Gradientenabstiegsverfahren gegangen wird und somit ist sie auch verantwortlich für das Ergebnis mit dem bestimmten Parameterraum im Minimum.

#### 4.5 

Wird Feature-Scaling angewandt, erhält man andere Werte für die Parameter, als ohne Feature-Scaling. 

**Solution**

Ja, weil mit feature scaling zum Beispiel vermieden werden kann in einem lokalen Optimum stehen zu bleiben.

In [6]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für Aufgabe 
aufgabe = 5
aufgaben_punkte[aufgabe -1] = 3

### Aufgabe 5: Lineare Separabilität

Gegeben ist folgendes Klassifikationsproblem:

Es gibt nur zwei binäre Features $x_1$ und $x_2$ mit $x_i \in \{0,1\}$.

Für den Zielwert gilt: $y \in \{0, 1\}$.

Dabei gilt folgende Klassenzuordnung $\vec{x} \rightarrow y$:
- $(0,0)\rightarrow 0$
- $(0,1)\rightarrow 1$
- $(1,0)\rightarrow 1$
- $(1,1)\rightarrow 0$

Ist dieses Problem ein linear separables Klassifikationsproblem. Begründen Sie Ihre Entscheidung.



Ihre Lösung hier oder auf einem separatem Blatt

Ja, lineare separabilität ist die Trennung von zwei Mengen mit einer linearen Funktion. Beim gegebenen Beispiel bilden sich genau zwei Mengen ab, die linear separiert werden können.

Eine Menge die $0$ abbildet:

- $(0,0)\rightarrow 0$
- $(1,1)\rightarrow 0$

Und eine Menge die $1$ abbildet:

- $(0,1)\rightarrow 1$
- $(1,0)\rightarrow 1$


In [7]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für Aufgabe 
aufgabe = 6
aufgaben_punkte[aufgabe -1] = 5

### Aufgabe 6: Kostenfunktion für log. Regression

Gegeben sind folgende Trainingsdaten für die logistische Regression:

$$
\mathcal D_{train} = 
\left\{ 
  (\vec x^{(1)}, y^{(1)}), 
  (\vec x^{(2)}, y^{(2)}), 
  (\vec x^{(3)}, y^{(3)}),
  (\vec x^{(4)}, y^{(4)}) 
\right\}
$$

mit 
- $(\vec x^{(1)}, y^{(1)}) = \left( (3.7, 2.1), 1 \right)$
- $(\vec x^{(2)}, y^{(2)}) = \left( (1.1, 1.2), 0 \right)$
- $(\vec x^{(3)}, y^{(3)}) = \left( (4.0, 2.0), 1 \right)$
- $(\vec x^{(4)}, y^{(4)}) = \left( (0.9, 1.4), 0 \right)$


mit der Hypothese für den "aktuellen" Parametervektor 
$\vec \theta= (-3., 1.1, 1.2)$. Dies ergibt folgende Hypothesenwerte

- $h^{(1)}(\vec x^{(1)}) = 0.973$
- $h^{(2)}(\vec x^{(2)}) = 0.413$
- $h^{(3)}(\vec x^{(3)}) = 0.978$
- $h^{(4)}(\vec x^{(4)}) = 0.418$ 


Wie berechnet sich die typische Kostenfunktion der logistischen Regression für die oben genannten Werte? Vereinfachen Sie den Ausdruck so weit wie möglich.




In [15]:
-1/4*((1*np.log2(0.973)+(1-1)*np.log2(1-0.973))+(0*np.log2(0.413)+(1-0)*np.log2(1-0.413))+(1*np.log2(0.978)+(1-1)*np.log2(1-0.978))+(0*np.log2(0.418)+(1-0)*np.log2(1-0.418)))

0.4052646132241667

**Solution**

$$- \frac{1}{m}  \sum_{i=1}^{m} 
    \left[  y^{(i)} \log h_\theta({\vec x}^{(i)})+
      (1 - y^{(i)}) \log \left( 1- h_\theta({\vec x}^{(i)})\right) \right] = -\frac{1}{4}\cdot \bigg( (1 \cdot \log(0.973)+ (1-1)\cdot \log(1-0.973)\bigg) +\bigg((0 \cdot \log(0.413)+ (1-0)\cdot \log(1-0.413)\bigg)+\bigg((1 \cdot \log(0.978)+ (1-1)\cdot \log(1-0.978)\bigg)+\bigg((0 \cdot \log(0.418)+ (1-0)\cdot \log(1-0.418)\bigg)=0.4052646132241667$$

In [8]:
##Ändern Sie nicht diese Zelle 

#(Max.) Punkte für Aufgabe 
aufgabe = 7
aufgaben_punkte[aufgabe -1] = 12

### Aufgabe 7. Entropie

Gegeben ist folgende Split-Tabelle für ein Feature mit dem Schwellenwert $v$:

|   | feature $\geq v$  | feature < $v$ |
|---|---|---|
| Klasse 0  | 44  | 7  |
| Klasse 1  | 12  |  88 |


Nomenklatur:
- Die Variable für die Klasse ist $Y$.
- Die Variable für den Spilt mit dem Feature ist $X$.

#### Aufgabe 7.1
Berechnen Sie die Entropie vor dem Split mit dem Feature $H(Y)$:

In [45]:
eBefore = -(51/100) * np.log2(51/151) - (100/151) * np.log2(100/151)

**Solution**

$$H(Y)=-\frac{51}{151} \cdot \log_2(\frac{51}{151})-\frac{100}{151}\cdot \log_2(\frac{100}{151})=1.1923902539419529$$

#### Aufgabe 7.2
Berechnen Sie die bedingte Entropie nach dem Split mit dem Feature, d.h. $H(Y \mid X)$

**Solution**

$$E_{smaller}=-(\frac{7}{95} \cdot \log_2(\frac{7}{95})+\frac{88}{95}\cdot \log_2(\frac{88}{95})=0.3795243778626806$$

$$E_{bigger}=-(\frac{44}{56} \cdot \log_2(\frac{44}{56})+\frac{12}{56}\cdot \log_2(\frac{12}{56})=0.74959525725948$$

$$H(Y|X)=\frac{95}{151} \cdot E_{smaller} + \frac{56}{151}\cdot E_{bigger}=0.5167692073078513$$

In [44]:
samples = 44 + 7 + 12 + 88
entropy_feature_smaller_v = -(7/95*np.log2(7/95)+88/95*np.log2(88/95))
entropy_feature_bigger_v = -(44/56*np.log2(44/56)+12/56*np.log2(12/56))
print(entropy_feature_smaller_v)
print(entropy_feature_bigger_v)
condP = 95./(samples) * entropy_feature_smaller_v + 56./(samples) * entropy_feature_bigger_v 
print(condP)

0.3795243778626806
0.74959525725948
0.5167692073078513


#### Aufgabe 7.3

Berechen Sie den Information Gain für den Split mit dem Feature.



**Solution**

$$I = H(Y)-H(Y|X) = 0.6756210466341016$$

In [46]:
iGain = eBefore-condP
print(iGain)

0.6756210466341016


------------------

#### Ende der Klausur 

In [10]:

# Maximale Punktzahl
print(aufgaben_punkte)

print("\nPunkte für Aufgaben:", aufgaben_punkte.sum())


[14.  5.  4. 10.  3.  5. 12.]

Punkte für Aufgaben: 53.0
